File: client/android/subak/app/src/test/java/com/subakstudio/subak/api/SubakApiTest.java
package com.subakstudio.subak.api;

import org.junit.Before;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

import rx.Subscriber;

import static org.junit.Assert.assertTrue;

/**
 * Created by jinwoomin on 8/6/16.
 */

public class SubakApiTest {

    private boolean taskDone;

    @Before
    public void setup() {
        taskDone = false;
    }

    private void waitUntilTaskDone() throws InterruptedException {
        while (!taskDone) {
            Thread.sleep(100);
        }
    }

    @Test
    public void testGetEngineList() throws Exception {
        SubakClient client = new SubakClient("http://localhost:8081");
        final List<Engine> result = new ArrayList<>();
        client.getEngineList()
                .subscribe(new Subscriber<Engine>() {
                    @Override
                    public void onCompleted() {
                        System.out.println("oncompleted");
                        taskDone = true;
                    }

                    @Override
                    public void onError(Throwable e) {
                        e.printStackTrace();
                        System.out.println("onerror: e=" + e);
                        taskDone = true;
                    }

                    @Override
                    public void onNext(Engine engine) {
                        System.out.println("onnext: engine=" + engine);
                        result.add(engine);
                    }
                });

        waitUntilTaskDone();

        assertTrue(result.size() > 0);
    }

    @Test
    public void testTrackList() throws Exception {
        final List<Track> result = new ArrayList<>();

        SubakClient client = new SubakClient("http://localhost:8081");
        Engine engine = new Engine();
        engine.setType(SubakApiInterface.ENGINE_TYPE_SEARCH);
        engine.setPath("/api/bobborst/year/billboard/top/year/:keyword");
        client.getTrackList(engine, "2016")
                .subscribe(new Subscriber<TrackListResponse>() {
                    @Override
                    public void onCompleted() {
                        System.out.println("oncompleted");
                        taskDone = true;
                    }

                    @Override
                    public void onError(Throwable e) {
                        System.out.println("onerror: e=" + e);
                        taskDone = true;
                    }

                    @Override
                    public void onNext(TrackListResponse trackListResponse) {
                        System.out.println("onnext: response=" + trackListResponse);
                        result.addAll(trackListResponse.getTracks());
                    }
                });

        waitUntilTaskDone();

        assertTrue(result.size() > 0);
    }
}

File: client/android/subak/app/src/test/java/com/subakstudio/subak/api/SubakApiTest.kt
package com.subakstudio.subak.api

import org.junit.Before
import org.junit.Test

import java.util.ArrayList

import rx.Subscriber

import org.junit.Assert.assertTrue

/**
 * Created by jinwoomin on 8/6/16.
 */

class SubakApiTest {

    private var taskDone: Boolean = false

    @Before
    fun setup() {
        taskDone = false
    }

    @Throws(InterruptedException::class)
    private fun waitUntilTaskDone() {
        while (!taskDone) {
            Thread.sleep(100)
        }
    }

    @Test
    @Throws(Exception::class)
    fun testGetEngineList() {
        val client = SubakClient("http://localhost:8081")
        val result = ArrayList<Engine>()
        client.engineList
                .subscribe(object : Subscriber<Engine>() {
                    override fun onCompleted() {
                        println("oncompleted")
                        taskDone = true
                    }

                    override fun onError(e: Throwable) {
                        e.printStackTrace()
                        println("onerror: e=" + e)
                        taskDone = true
                    }

                    override fun onNext(engine: Engine) {
                        println("onnext: engine=" + engine)
                        result.add(engine)
                    }
                })

        waitUntilTaskDone()

        assertTrue(result.size > 0)
    }

    @Test
    @Throws(Exception::class)
    fun testTrackList() {
        val result = ArrayList<Track>()

        val client = SubakClient("http://localhost:8081")
        val engine = Engine()
        engine.type = SubakApiInterface.ENGINE_TYPE_SEARCH
        engine.path = "/api/bobborst/year/billboard/top/year/:keyword"
        client.getTrackList(engine, "2016")
                .subscribe(object : Subscriber<TrackListResponse>() {
                    override fun onCompleted() {
                        println("oncompleted")
                        taskDone = true
                    }

                    override fun onError(e: Throwable) {
                        println("onerror: e=" + e)
                        taskDone = true
                    }

                    override fun onNext(trackListResponse: TrackListResponse) {
                        println("onnext: response=" + trackListResponse)
                        result.addAll(trackListResponse.tracks)
                    }
                })

        waitUntilTaskDone()

        assertTrue(result.size > 0)
    }
}
